### PR TITLE
[FRONT-97] Filter Columns

### PR DESCRIPTION
**Pull request by ArnoClaude**
## Description of Issue
[Linear Task](https://linear.app/la-famiglia-project/issue/FRONT-97/fix-and-finish-table-filtering-and-sorting)

## How I implemented
I orientated myself on the Notion table filtering a lot. They also just allow one filter per column for basic filtering, which came in handy as it made the implementation easier (can't remember exactly what the problem was that I encountered, something about updating the filter array state being easier when only enabling one filter per column).
Let me know if having just one filter per column is a big hindrance, and I'll give it another try to enable multiple filters per column :)

## Other
Todo:
- Remove the blue outline from modals (just check what I did for the sidebar profile modal)

### CODE REVIEW

File: src/components/pure/ProjectsTable/columns.tsx
Comment (kigadye) [inline src/components/pure/ProjectsTable/columns.tsx:1]: Please revert to contrib., I know you have a huge screen but most people will use small laptops where it looks kinda ugly if the column title is too big

**Review by kigadye — CHANGES_REQUESTED** [review]
Please remove outline on modals similar to updated sidebar/ profile modal + fix padding on the right:
Screenshot 2023-06-11 at 15 37 55

Make z-index of topbar greater than that of sidebar or align the modal to the left of the button like the (would prefer the latter, same for filter modal):
Screenshot 2023-06-11 at 15 33 12

The edit columns modal now doesn't close when you click outside any more.

File: src/components/pure/ProjectsTable/columns.tsx
Comment (ArnoClaude) [reply src/components/pure/ProjectsTable/columns.tsx:1]: Reverted to `Contrib.`.
Changed z-indices for proper dropdown overlaying.
Regarding the outline, right-padding, and closing modals when clicking outside, I'll tackle that later today, I couldn't get it to work properly just now :)

**Comment by DamianGrosz2** [comment]
Bildschirm­foto 2023-06-13 um 12 30 22
- The box could have some more padding.

- This arrow down is a bit unintuitive 
I think ">" and "<" will be enough, the other options would probably not be used anyway. 
So we could do: Stars "greater" 2000 / Stars "less than" 2000

This would solve the arrow problem.

**Comment by DamianGrosz2** [comment]
In terms of usability: we could have the delete button outside of the modal, so that it's always visible. Not a must though, what do you think?

**Comment by DamianGrosz2** [comment]
I think having a filter for both min and max of a column would be useful. If that's out of scope no worries, then let's add it to improvements.

Other than that, great job!

**Comment by ArnoClaude** [comment]
> Bildschirm­foto 2023-06-13 um 12 30 22
> * The box could have some more padding.
> * This arrow down is a bit unintuitive
>   I think ">" and "<" will be enough, the other options would probably not be used anyway.
>   So we could do: Stars "greater" 2000 / Stars "less than" 2000
> 
> This would solve the arrow problem.

Agree on the padding. But for the arrow dropdown; Notion does the same. I reckon if Notion does it that way, it probably is fine in terms of UX. What do you think about this? @kigadye 

image

**Comment by DamianGrosz2** [comment]
> > Bildschirm­foto 2023-06-13 um 12 30 22
> > * The box could have some more padding.
> > * This arrow down is a bit unintuitive
> >   I think ">" and "<" will be enough, the other options would probably not be used anyway.
> >   So we could do: Stars "greater" 2000 / Stars "less than" 2000
> > This would solve the arrow problem.
> 
> Agree on the padding. But for the arrow dropdown; Notion does the same. I reckon if Notion does it that way, it probably is fine in terms of UX. What do you think about this? @kigadye
> 
> image

Ok makes sense, then how about writing the name "greater/less" then the arrow makes sense. Otherwise imo arrow left + arrow down is a bit confusing. @ArnoClaude

**Comment by ArnoClaude** [comment]
> > > Bildschirm­foto 2023-06-13 um 12 30 22
> > > * The box could have some more padding.
> > > * This arrow down is a bit unintuitive
> > >   I think ">" and "<" will be enough, the other options would probably not be used anyway.
> > >   So we could do: Stars "greater" 2000 / Stars "less than" 2000
> > > This would solve the arrow problem.
> > 
> > 
> > Agree on the padding. But for the arrow dropdown; Notion does the same. I reckon if Notion does it that way, it probably is fine in terms of UX. What do you think about this? @kigadye
> > image
> 
> Ok makes sense, then how about writing the name "greater/less" then the arrow makes sense. Otherwise imo arrow left + arrow down is a bit confusing. @ArnoClaude

Even for numeric operators, Notion does it that way :D But sure, I can change that :)
image

File: src/components/page/overview/FilterBar.tsx [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import { TableFilter } from '@/components/page/overview/TableFilter'
+import FilterFormModal from '@/components/page/overview/FilterFormModal'
+
+type FilterBarProps = {
+  filters: TableFilter[]
+  removeFilter: (filter: TableFilter) => void
+  updateFilter(filter: TableFilter): void
+}
```
Comment (kigadye) [inline src/components/page/overview/FilterBar.tsx:8]: Add newline

File: src/components/page/overview/FilterBar.tsx [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  <div className="flex flex-row justify-between border-b border-gray-800 px-6 py-2.5">
+    <div className="flex flex-row gap-3">
+      {filters.map((filter: TableFilter) => (
+        <FilterFormModal filter={filter} removeFilter={removeFilter} updateFilter={updateFilter} />
+      ))}
+    </div>
+
+    <div className="flex flex-row items-center">
+      <p className="text-14">Showing 10&nbsp;</p>
+      <p className="text-14 text-gray-500">/&nbsp;25 entries</p>
```
Comment (kigadye) [inline src/components/page/overview/FilterBar.tsx:19]: Make values dynamic (just pass initial & current array lengths down as props)

File: src/components/page/overview/Table.tsx [hunk 3]
```diff
@@ -1,15 +1,71 @@
...
+  const cellValue = cell.getValue()
+  const { operator, value } = filter
+  if (
+    value !== undefined &&
+    (operator === 'is' ||
+      operator === 'is not' ||
+      operator === 'contains' ||
+      operator === 'does not contain' ||
+      operator === 'starts with' ||
+      operator === 'ends with')
```
Comment (kigadye) [inline src/components/page/overview/Table.tsx:25]: You can do `Object.values(enum).includes(operator)` instead of this

**Review by kigadye — DISMISSED** [review]
Great work overall! Filtering works perfectly. A few code quality and UI improvements are possible though:
- Close modals when clicking outside of them, I believe this was the case before you changed the edit columns modal, so the easiest way would be to undo whatever you did to that, I think the UX isn't too bad if it closes after each click (this should also solve the issue of having 2 models open at the same time)
- Remove the blue outline from modals (just check what I did for the sidebar profile modal)
- Add padding to filter modal and change background color of input to dark grey
- Fix padding on the right of items in filters modal
- Change the options of the timeframe button to today, this week and this month to reflect the GitHub trending categories and hide it on the compare page
- Rename the sorts button to sorting

Screenshot 2023-06-15 at 01 36 42

File: src/components/page/overview/Table.tsx [hunk 4]
```diff
@@ -1,15 +1,71 @@
...
+  ) {
+    const cellValueStr = cellValue as string
+    switch (operator) {
+      case StringTableFilterOperator.IS:
```
[the remark this answers is not in the document]
Comment (ArnoClaude) [reply src/components/page/overview/Table.tsx:29]: I prefer long and descriptive names in general. Everyone uses autocomplete anyways.

**Review by DamianGrosz2 — CHANGES_REQUESTED** [review]
Great job Arno! Overall, looks good to me now, though:
I'm not sure if anybody will ever use not equals as a filter.
Imo we could remove that, and the equals filter aswell?

Also I really think that > is sufficient, why would you also need >=?
That gives the user an unnecessary decision and worsens the UX.

**Comment by ArnoClaude** [comment]
> Great job Arno! Overall, looks good to me now, though: I'm not sure if anybody will ever use not equals as a filter. Imo we could remove that, and the equals filter aswell?
> 
> Also I really think that > is sufficient, why would you also need >=? That gives the user an unnecessary decision and worsens the UX.

Sure :)

**Review by DamianGrosz2 — APPROVED** [review]
Great

**Review by kigadye — APPROVED** [review]

**Pull request merged**